src/shims/unix/linux/eventfd.rs

```diff
@@ -217,8 +179,195 @@ pub trait EvalContextExt<'tcx>: crate::MiriInterpCxExt<'tcx> {
+fn blocking_eventfd_write_callback<'tcx>(
+    num: u64,
+    buf_place: &MPlaceTy<'tcx>,
+    dest: &MPlaceTy<'tcx>,
+    weak_eventfd: WeakFileDescriptionRef,
+    ecx: &mut MiriInterpCx<'tcx>,
+) -> InterpResult<'tcx> {
+    let Some(eventfd_ref) = weak_eventfd.upgrade() else {
+        throw_unsup_format!("eventfd FD got closed while blocking.")
+    };
+
+    // Since we pass the weak file description ref to the callback function, it is guaranteed to be
+    // an eventfd file description.
+    let eventfd = eventfd_ref.downcast::<Event>().unwrap();
+
+    // Future `read` calls will synchronize with this write, so update the FD clock.
+    ecx.release_clock(|clock| {
+        eventfd.clock.borrow_mut().join(clock);
+    });
+
+    // In the happy case, the new_count is checked before executing this callback.
+    // In the case where the counter previously overflows or has the value u64::MAX,
+    // the counter will be set to 0 before the callback is executed.
+    let new_count = eventfd.counter.get().checked_add(num).unwrap();
+    eventfd.counter.set(new_count);
+
+    // When any of the event happened, we check and update the status of all supported event
+    // types for current file description.
+    ecx.check_and_update_readiness(&eventfd_ref)?;
+
+    // Unblock *all* threads previously blocked on `read`.
+    let mut blocked_read_tid = eventfd.blocked_read_tid.borrow_mut();
+    while let Some(tid) = blocked_read_tid.pop() {
+        ecx.unblock_thread(tid, BlockReason::Eventfd)?;
+    }
+
+    // Return how many bytes we read.
+    ecx.write_int(buf_place.layout.size.bytes(), dest)
+}
+
+/// Check the value in write then block thread if necessary.
+/// `reblock` records if this thread is blocked for the first time.
+fn check_write_value_and_block_thread<'tcx>(
+    num: u64,
+    buf_place: MPlaceTy<'tcx>,
+    dest: &MPlaceTy<'tcx>,
+    weak_eventfd: WeakFileDescriptionRef,
+    reblock: bool,
+    ecx: &mut MiriInterpCx<'tcx>,
+) -> InterpResult<'tcx> {
+    let Some(eventfd_ref) = weak_eventfd.upgrade() else {
+        throw_unsup_format!("eventfd FD got closed while blocking.")
+    };
+
+    // Since we pass the weak file description ref, it is guaranteed to be
+    // an eventfd file description.
+    let eventfd = eventfd_ref.downcast::<Event>().unwrap();
+
+    match eventfd.counter.get().checked_add(num) {
+        Some(_new_count @ 0..=MAX_COUNTER) => {
+            return blocking_eventfd_write_callback(num, &buf_place, dest, weak_eventfd, ecx);
+        }
+        None | Some(u64::MAX) => {
+            if eventfd.is_nonblock {
+                return ecx.set_last_error_and_return(ErrorKind::WouldBlock, dest);
+            }
+
+            let dest = dest.clone();
+
+            // We don't want to record the same thread if the current thread is still blocked.
+            if !reblock {
+                eventfd.blocked_write_tid.borrow_mut().push(ecx.active_thread());
+            }
```

Comment:
If this function is recursively called, it will run into ``already borrowed: BorrowMutError`` panic and dropping ``blocked_write_tid`` like below won't work. 

```rust
let mut blocked_write_tid = eventfd.blocked_write_tid.borrow_mut();
blocked_write_tid.push(ecx.active_thread());
drop(blocked_write_tid)
```

So I use this to prevent reborrow, and prevent a thread id being pushed to ``blocked_write_tid`` again if it is already there.

Reply:
That sounds like somewhere, a `RefMut` should be dropped earlier.

Reply:
Yea, I will check again and see how can I prevent the panic. I don't really like the idea of adding ``reblock``.

Reply:
Yea, the issue is indeed somewhere else.